### ISSUE TITLE
Missing platform ability: calling the original value of built-in functions

### ISSUE THREAD

**Issue by domenic** [drufball/layered-apis#6]
So that multiple implementations of a layered API can exist, layered API specifications should generally call the underlying operation, not the globally-modifiable public API.

For example, as of the time of this writing the [async local storage](https://domenic.github.io/async-local-storage/) spec says

> Let request be the result of performing the steps listed in the description of IDBObjectStore's get() method on _store_, given the argument _key_.

instead of saying to call `store.get(key)`. It also says

> If IsArray(value) is true, return true.

instead of asking if `Array.isArray(value)` is true.

This is motivated by a desire to allow implementations more internal flexibility in how they implement the specification, without their every move being interceptable by JavaScript. For example, they could implement the specification in C++, calling in to their existing Web IDL methods.

However, this is not a capability that is currently available to web developers. It thus violates the layered API constraint of not using any un-layered magic.

This issue is for discussing potential resolutions to this contradiction.

**Comment by domenic** [comment]
One solution that seems relatively simple API-wise is `self.getOriginalPropertyDescriptor()`, which you would use like so:

```js
// Save some useful stuff
const Reflect_call = self.getOriginalPropertyDescriptor("Reflect.call").value;
const Array_isArray = self.getOriginalPropertyDescriptor("Array.isArray").value;
const IDBObjectStore_get = self.getOriginalPropertyDescriptor("IDBObjectStore.prototype.get").value;
const IDBRequest_result = self.getOriginalPropertyDescriptor("IDBRequest.prototype.result").get;

// Functions:
const shouldBeTrue = Array_isArray([1, 2, 3]);

// Methods:
const request = Reflect_call(IDBObjectStore_get, someStore, [someQuery]);

// Getters:
const theResult = Reflect_call(IDBRequest_result, request);
```

However, I'm not sure how feasible this would be to implement. In particular, the string -> original property mapping seems a bit tricky. I think it is probably doable if we limited it to all "globals" and their "prototype" properties. ("Globals" here = exposed through Web IDL, or defined as a global property in the JS spec.)

**Comment by bmeck** [comment]
Some issues we need to cover:

1. How to deal with Symbols like getting the `Symbol.iterator` of various types.
2. Termination conditions. I think not walking getter/setters to add to the paths seems sane.
3. There are often multiple ways to get to a variety of things (e.g. you can get to Object.prototype from all sorts of locations). Is there only 1 way to get to any given primordial? If so, would breadth first search be the right way to approach which path "wins"?

**Comment by bmeck** [comment]
Draft/polyfill of a potential design : https://gist.github.com/bmeck/2a5e141cca17ecce7c252f5baa47eb48

* Does breadth first search except for some JS language builtins so that some things show up in a more logical order. 
  * Only retains a single path for value, will walk and iterate paths only for first visit.
* Handles symbols by changing the API signature to be `getOriginalPropertyDescriptor(...path : KeyLike)`.
* Freezes the descriptor but does not unlink from `Object.prototype` so some care still needs to be taken if people are testing for `value` vs `get` or `set` being on the descriptor.
* It does not visit `get`/`set` of property descriptors when forming paths. This would be problematic for paths since it branches a path in an odd way and does not seem to add value.

The polyfill has some taintability right now, but could be coded to not have that taintability if loaded first. This is more a research into the API usability / design constraints right now.

**Comment by ljharb** [comment]
I *love* the idea of getting "original" descriptors, but polyfills (and SES) all need to be able to redefine what "original" means - iow, i think intrinsics need to always be shadow-able by first-run code.

**Comment by domenic** [comment]
@ljharb this API is specifically intended for code that does not want to get interfered with by polyfill code, even if the polyfill code runs first. (For example, code that intends to work similar to how the browser works.) So, it would not be shadow-able in that sense.

This API, like many web APIs, probably would not be allowed in SES environments.

**Comment by bmeck** [comment]
This effectively grants access to primordials in a way that is not intended to be able to be intercepted. This would allow things like polyfill code to be loaded lazily instead of placing an up front cost as well. I'm not entirely sure of why this needs to allow replacement since it is meant to grant primordials that are provided by the host rather than by userland.

**Comment by domenic** [comment]
One interesting variant I came up with that generalizes this slightly to allow author code to also create its own set of original-stuff is the following:

```js
const o = { X, Y, Z };
const getMyOPDs = saveOriginalPropertyDescriptors(o);

getMyOPDs("X"); // returns original property descriptor on o for "X"
getMyOPDs("X", "prototype"); // returns the original property descriptor on o's original "X" for "prototype"
// ... etc ...
```

Then you'd conceptualize the global `self.getOriginalPropertyDescriptor(...args)` as if it had been created on browser startup by

```js
self.getOriginalPropertyDescriptor = saveOriginalPropertyDescriptors(self);
```

@bmeck's code has disappeared (I think codepen has a time limit?) but I imagine the implementation of `saveOriginalPropertyDescriptors` would be very similar to what I remember that code being like.

The global `self.getOriginalPropertyDescriptor` would probably not actually be implemented this way, as it should instead be lazy for memory usage and performance. But on the small scale of individual author libraries which want to be able to provide the same sort of guarantees, the crawling done by `saveOriginalPropertyDescriptors` seems good.

Note that we can also add `saveOriginalPropertyDescriptors` later, even if we start with just `self.getOriginalPropertyDescriptor`.

**Comment by bmeck** [comment]
@domenic updated so that it should have a permanent link.

**Comment by bzbarsky** [comment]
Some thoughts: 

1. Having non-configurable things on Window is a bit complicated, because that violates ES invariants.  See the longstanding HTML spec bugs about this...  It would really be good to sort those out before we add more unforgeable things on Window.
2. There are in fact multiple paths that can be used to get to the same object, especially if you want to walk prototype links.
3. `self` is not unforgeable (and in fact is replaceable), so there's not much point in making `saveOriginalPropertyDescriptors` unforgeable if you plan to call it via `self`.  Calling it via `window` would work, though....

**Comment by bmeck** [comment]
@bzbarsky I'm only going to talk about the multiple paths topic in this comment.

There are multiple ways for getting to references such as the global such as:

```mjs
window
window['self']
window['window']
```

For the case of my example implementation we are treating `window` as the root of our traversal. By doing a breadth first traversal we get the smallest path to a reference as the earliest point of traversal. 
By keeping the key that gets to a reference but stoping the traversal when a visited node is found, we keep both `getOriginalPropertyDescriptor('self')` and `getOriginalPropertyDescriptor('window')` but do not traverse into them. This means that other paths like `getOriginalPropertyDescriptor('self', 'self')` or other longer paths are not able to be used since `self` is only traversed once.

Another example is with `Location`. `getOriginalPropertyDescriptor('self', 'location')` is not preserved, but `getOriginalPropertyDescriptor('location')` is.

In summary, all paths to a reference are preserved but only the first path is traversed for subpaths. I don't know of any reference that has 2 properties at the same depth that cause a collision currently, but would be interested if there was an example. If so we could mark both as sharing the same traversal, which seems fine and still avoids infinite paths.

**Comment by domenic** [comment]
I think the MVP version of this just includes all [Exposed] globals, their `prototype` properties, and any statics/constants on them. That gets you quite far; at least, for the purposes of writing robust code, it might be all you need. We could certainly expand it to a full traversal later if we find something important was missing.

Will try to sort out the ES invariants violation; IIRC we had rough TC39 consensus on adding a hook to allow Window to be special, but we never quite landed it in the spec.

Good point on the examples should not be using `self` :). I try to be more explicit in general about global properties, and I prefer `self` to `window` since it works in workers too, but in this specialized case it's clearly not the right move.

**Comment by bzbarsky** [comment]
> I think the MVP version of this just includes all [Exposed] globals, their prototype properties, and any statics/constants on them.

By "prototype" properties I assume you mean "own properties of Foo.prototype"?  If so, I agree.  I'm OK with making the API conceptually extendable to other things in the future, but I think we should start with this set.

That said, given the current API shape ther are going to be some interop problems for a bit because of the unresolved Document/HTMLDocument situation and the fact that UAs have properties on different prototypes as a result.  :(  There will also be interop problems any time a property moves to a different prototype.

Given that web authors generally don't need to worry about where on the proto chain properties live, it might be nice if they could continue not caring with this API.  That would argue for an API shape where you pass in an object and a property name and the browser returns the "canonical" property descriptor for that property name on that object (internally doing a walk over the canonical prototype chain, etc).

That won't work for statics, but arguably those are already a bit different.  That is, it might make sense to do:

    window.getWhateverWeNameThis("InterfaceName", "staticName");

and

    window.getWhateverWeNameThisOtherThing(thisObj, "propName");

One other question, on the `self` vs `window` topic.  What about workers?   Do we want to expose this API in workers?  Can we make `self` unforgeable in workers or something?  How would code detect whether it should use `window` or `self`?

**Comment by domenic** [comment]
> By "prototype" properties I assume you mean "own properties of Foo.prototype"? If so, I agree. I'm OK with making the API conceptually extendable to other things in the future, but I think we should start with this set.

Yep, sorry that I was imprecise. Trying to be more precise, I think the MVP is:

1. Constructors
   - All [Exposed] Web IDL-defined interfaces
   - All properties listed in https://tc39.github.io/ecma262/#sec-constructor-properties-of-the-global-object
2. Namespaces
   - All properties listed in https://tc39.github.io/ecma262/#sec-other-properties-of-the-global-object
   - All [Exposed] Web IDL-defined namespaces
3. All `prototype` properties of (1) and (2)
4. All own properties of (1), (2), and (3)

(4) will need a bit more precision, e.g. I think we don't need `length`, `constructor`, or `name`. If we worked from ES including them would be natural but if we worked from Web IDL it would be less natural. I don't feel strongly either way about including or excluding them but we should make a decision.

> (musings on proto chains)

I like the direction this is heading. Maybe we should also split out the accessors vs. functions case. So I am thinking

```js
window.getOriginalStaticValue("InterfaceName", "staticName"); // returns value directly
window.getOriginalInstanceValue("InterfaceName", "methodOrValuePropertyName"); // returns value directly
window.getOriginalAccessor("InterfaceName", "accessorPropertyName"); // returns { get, set } object, maybe null-proto
```

You did `window.getWhateverWeNameThisOtherThing(thisObj, "propName")` but this seems trickier to implement than using `"InterfaceName"`; you'd have to go from object brand -> interface name -> internal lookup table, right? Thoughts?

> One other question, on the self vs window topic. What about workers? Do we want to expose this API in workers? Can we make self unforgeable in workers or something? How would code detect whether it should use window or self?

I think we definitely should support workers.

I might be missing something obvious, but if you just use bare `getOriginalThingy()`, won't that work?

**Comment by bzbarsky** [comment]
> you'd have to go from object brand -> interface name -> internal lookup table, right?

At least in Gecko, no.  In Gecko, we can go from the brand directly to the thing we care about.  Going from a string we'd have to look up the string in some lookup table that doesn't even exist at the moment to derive the object brand, etc.

> but if you just use bare getOriginalThingy(), won't that work?

Not if the worker imports scripts that use `var getOriginalThingy` or `let getOriginalThingy`, right?

**Comment by domenic** [comment]
I've started to dump all this into an actual proposal at https://github.com/domenic/get-originals. It's rather skeletal right now, as I need to wind down for the day. But I think captures a lot of the things we talked about.

As a summary, it looks like the following suffice:

- `getOriginalGlobal("globalProperty")`
- `getOriginalProperty(obj, "propertyName")`
- `setOriginalProperty(obj, "propertyName", newValue)`
- `callOriginalMethod(obj, "methodName", ...args)` (optional)

My biggest worry at this point is whether this instance-based approach is workable for JS built-ins (as opposed to Web IDL ones). I'll need to think harder about that over the weekend.

For now, feel free to continue discussion here, or open issues over there.

**Comment by domenic** [comment]
Status update: I fleshed out https://github.com/domenic/get-originals ([permalink to version as of the time of this writing](https://github.com/domenic/get-originals/tree/e3d9ea548503395189e13e375dfadfca09b07f8b)) with the above methods. But thinking about implementation strategies, those methods seems pretty hard to implement. E.g. making `getOriginalProperty()` work for statics, or saying that `getOriginalGlobal()` works for all properties of the global object, seems hard.

Tomorrow I'll try to update it based on an easier-to-implement idea. Preview/note to self:

```js
// ### Implement using a lookup table:

getOriginalConstructor("TypeError");

getOriginalConstructor("TypeError") === TypeError; // important so that instanceof works


// ### Re-create the JS function object

getOriginalStatic("Array", "isArray");
getOriginalStatic("JSON", "parse");
getOriginalStatic("Response", "redirect");

getOriginalStatic("JSON", "parse") !=== JSON.parse; // since it gets recreated


// ### Necessary for making getOriginalProperty useful:

originalSelf === self; // but, originalSelf is unforgeable


// ### Implement by:
//   - Unwrapping platform object and calling C++ method/getter/setter, OR
//   - Finding JS spec brand and calling appropriate original method/getter/setter code, hopefully in a fast way

getOriginalProperty(originalSelf, "document");
getOriginalProperty(document, "body");
setOriginalProperty(document, "cookie", "foo=bar");
callOriginalMethod(document.body, "setAttribute", "bgcolor", "red");
callOriginalMethod(self, "decodeURI", "abcd");


// ### Things that do not work, even if you might expect them to:

getOriginalProperty(self, "TypeError");           // use getOriginalConstructor()
getOriginalProperty(self, "Node");                // use getOriginalConstructor()
getOriginalProperty(Array, "isArray");            // use getOriginalStatic()
callOriginalMethod(Array, "isArray", [1, 2, 3]);  // use getOriginalStatic()
getOriginalProperty(self, "encodeURIComponent");  // use callOriginalMethod()
getOriginalProperty(self, "close");               // use callOriginalMethod()
getOriginalProperty(Node, "COMMENT_NODE");        // use the literal 8; no API for constants
getOriginalProperty(self, "NaN");                 // use 0/0 or similar; no API for constants
```

**Comment by domenic** [comment]
And, done: https://github.com/domenic/get-originals now has an API surface that seems much more implementable based on my (admittedly non-expert) knowledge.

I've asked for implementability feedback from Blink and V8 folks on https://groups.google.com/a/chromium.org/forum/#!topic/platform-architecture-dev/u3cLa-BXGzo if people want to follow along there, or feel free to chime in in the proposal repo.

Let me close this issue so we can redirect discussion to the proposal repo. I'll edit the OP accordingly.
